libs/base/Data/Bits.idr

```diff
@@ -0,0 +1,200 @@
+  shiftR x    = prim__shr_Bits16 x . fromInteger . cast . fst
+  shiftL x    = prim__shl_Bits16 x . fromInteger . cast . fst
+
+  popCount x0 =
+    -- see https://stackoverflow.com/questions/109023/how-to-count-the-number-of-set-bits-in-a-32-bit-integer
+    let x1 = (x0 .&. 0x5555) + ((x0 `shiftR` fromNat 1) .&. 0x5555)
+        x2 = (x1 .&. 0x3333) + ((x1 `shiftR` fromNat 2) .&. 0x3333)
+        x3 = ((x2 + (x2 `shiftR` fromNat 4)) .&. 0x0F0F)
+        x4 = (x3 * 0x0101) `shiftR` fromNat 8
+     in fromInteger $ cast x4
+
+public export %inline
+Bits Bits32 where
+  Index       = Subset Nat (`LT` 32)
+  (.&.)       = prim__and_Bits32
+  (.|.)       = prim__or_Bits32
+  xor         = prim__xor_Bits32
+  bit         = (1 `shiftL`)
+  zeroBits    = 0
+  testBit x i = (x .&. bit i) /= 0
+  shiftR x    = prim__shr_Bits32 x . fromInteger . cast . fst
+  shiftL x    = prim__shl_Bits32 x . fromInteger . cast . fst
+
+  popCount x0 =
+    -- see https://stackoverflow.com/questions/109023/how-to-count-the-number-of-set-bits-in-a-32-bit-integer
+    let x1 = (x0 .&. 0x55555555) + ((x0 `shiftR` fromNat 1) .&. 0x55555555)
+        x2 = (x1 .&. 0x33333333) + ((x1 `shiftR` fromNat 2) .&. 0x33333333)
+        x3 = ((x2 + (x2 `shiftR` fromNat 4)) .&. 0x0F0F0F0F)
+        x4 = (x3 * 0x01010101) `shiftR` fromNat 24
+     in fromInteger $ cast x4
+
+public export %inline
+Bits Int where
+  Index       = Subset Nat (`LT` 64)
+  (.&.)       = prim__and_Int
+  (.|.)       = prim__or_Int
+  xor         = prim__xor_Int
+  bit         = (1 `shiftL`)
+  zeroBits    = 0
+  testBit x i = (x .&. bit i) /= 0
+  shiftR x    = prim__shr_Int x . cast . fst
+  shiftL x    = prim__shl_Int x . cast . fst
+
+  popCount x =
+    -- see https://stackoverflow.com/questions/109023/how-to-count-the-number-of-set-bits-in-a-32-bit-integer
+    -- We have to treat negative numbers separately in order to
+    -- prevent overflows in the first addition.
+    -- The top bit is therefore cleared and 1 is added in the end
+    -- in case of a negative number
+    let x0 = x `clearBit` fromNat 63
+        x1 = (x0 .&. 0x5555555555555555)
+           + ((x0 `shiftR` fromNat 1) .&. 0x5555555555555555)
+        x2 = (x1 .&. 0x3333333333333333)
+           + ((x1 `shiftR` fromNat 2) .&. 0x3333333333333333)
+        x3 = ((x2 + (x2 `shiftR` fromNat 4)) .&. 0x0F0F0F0F0F0F0F0F)
+        x4 = (x3 * 0x0101010101010101) `shiftR` fromNat 56
+        x5 = if x < 0 then x4 + 1 else x4
+     in fromInteger $ cast x5
+
+--------------------------------------------------------------------------------
+--          FiniteBits
+--------------------------------------------------------------------------------
+
+public export
+interface Bits a => FiniteBits a where
+  ||| Return the number of bits in values of type `t`.
+  bitSize : Nat
+
+  ||| Properly correlates `bitSize` and `Index`.
+  bitsToIndex : Fin bitSize -> Index {a}
+
+  ||| Returns the bitwise complement of a value.
+  complement : a -> a
```

Comment:
Does `complement` belong in `FiniteBits`? `Integer` isn't `FiniteBits` yet

```
complement : Integer -> Integer
complement x = xor x -1
```

just like all the other implementations. Chez scheme `logxor` supports bignums at least. The missing piece is a `oneBits` in `Bits`

Reply:
I assume with `oneBits` you mean the value with the first bit set to 1?

Reply:
No, all bits set to one. In twos complement that's -1, as in all the current implementations

Reply:
oneBits = complement zeroBits

Reply:
I moved `complement` to Data.Bits and added `oneBits`. In addition, I moved `popCount` to `FiniteBits` as it makes more sense there. I just quickly checked with Haskell, where `popCount` is defined on `Bits`: There, `popCount ((-1) :: Integer)` returns -1, probably not what we want.

Reply:
👍

Just for completeness, complement in Bits prevents an instance for "unsigned bignums" e.g. Nat, but I don't think that instance is a good idea anyway, and adding a third class is not warranted here imo.